cpp/oneapi/dal/algo/kmeans/backend/gpu/update_clusters.hpp

```diff
@@ -0,0 +1,104 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+#include "oneapi/dal/algo/kmeans/backend/gpu/kernels_integral.hpp"
+#include "oneapi/dal/algo/kmeans/backend/gpu/kernels_fp.hpp"
+
+namespace oneapi::dal::kmeans::backend {
+
+namespace pr = dal::backend::primitives;
+namespace bk = dal::backend;
+
+template <typename Float>
+auto update_clusters(sycl::queue& queue,
+                     std::int64_t cluster_count,
+                     std::int64_t max_iteration_count,
+                     double accuracy_threshold,
+                     std::int64_t block_rows,
+                     std::int64_t part_count,
+                     const pr::ndarray<Float, 2> data,
+                     const pr::ndarray<Float, 2> initial_centroids,
+                     pr::ndarray<Float, 2> centroids,
+                     pr::ndarray<Float, 2> partial_centroids,
+                     pr::ndarray<Float, 2> distance_block,
+                     pr::ndarray<Float, 2> closest_distances,
+                     pr::ndarray<Float, 1> objective_function,
+                     pr::ndarray<std::int32_t, 2> labels,
+                     pr::ndarray<std::int32_t, 1> counters,
+                     pr::ndarray<std::int32_t, 1> candidate_indices,
+                     pr::ndarray<Float, 1> candidate_distances,
+                     pr::ndarray<std::int32_t, 1> empty_cluster_count,
+                     const bk::event_vector& deps = {}) {
```

Comment:
This is not acceptable, I suggest to introduce a class with the state, such as `cluster_updater`. It could allocate temporal arrays and provide getters to them + implement the `update` function.

Reply:
Fixed